q2_fmt/_peds.py

```diff
@@ -468,3 +484,89 @@ def _create_masking(time_metadata, donor_df, recip_df, reference_column):
 def _mask_recipient(donor_mask, recip_df):
     maskedrecip = donor_mask & recip_df
     return maskedrecip
+
+
+def peds_bootstrap(table: pd.DataFrame, metadata: qiime2.Metadata,
+                   time_column: str, reference_column: str,
+                   subject_column: str,
+                   filter_missing_references: bool = False,
+                   drop_incomplete_subjects: bool = False,
+                   drop_incomplete_timepoint: list = None,
+                   bootstrap_replicates: int = 999) -> (pd.DataFrame):
+
+    metadata_df = metadata.to_dataframe()
+    donor = metadata_df[metadata_df.index.isin(
+        _check_reference_column(metadata=metadata_df,
+                                reference_column=reference_column))]
+
+    recipient = metadata_df.loc[metadata_df[reference_column].notnull()]
+    fake_donor = pd.DataFrame([])
+    for i in range(0, bootstrap_replicates+1):
+        if i == 0:
+            peds = sample_peds(
+                table=table, metadata=metadata,
+                time_column=time_column,
+                reference_column=reference_column,
+                subject_column=subject_column,
+                filter_missing_references=filter_missing_references,
+                drop_incomplete_subjects=drop_incomplete_subjects,
+                drop_incomplete_timepoint=drop_incomplete_timepoint)
+            real_temp = peds["measure"]
+        else:
+            shifted_list = recipient[reference_column].sample(frac=1).to_list()
+            recipient.loc[:, reference_column] = shifted_list
+            metadata_df = pd.concat([donor, recipient])
+            metadata = qiime2.Metadata(metadata_df)
```

Comment:
I think you should move your metadata shuffling out to a new private function (e.g., `_shuffle_donor_associations`), that will be easy to directly test and will consolidate that logic. 

Also, we should be thinking about how to handle the case where there aren't enough donors for this to be a meaningful shuffle (e.g., if there is only one donor). If we have special handling, that should early on in `peds_bootstrap` (so we don't do the check on each iteration, for example). It should be the case where we just don't get a significant result in that instance, but can you confirm that, and then also do some tests to get a feel for what might make sense here?

Reply:
I can do this but I think that your above suggestionhas turned this into 2 lines of code: 

`recipient[reference_column].sample(frac=1).reset_index(drop=True)
metadata = qiime2.Metadata(metadata_df)
`
Should that still be a seperate function? I can see the value of being able to direct test that even if its less lines of code.

Reply:
Also re: not enough donors.

This should be addressed by the new statistics I am going to add. I do plan to simulate data to define a threshold for how many donors we need.

Reply:
I have looked at this with Evan and I believe that the only issue is if we aren't able to shuffle(ie only one donor) . If we have more than one this will action will behave rationally. 
I have added an AssertionError if only one donor or recipient was added.